### PR TITLE
feat(node-v21): added support for node v21

### PR DESCRIPTION
**Pull request by aryamohanan**
Problems:
Node v21 is not working with ESM https://github.com/nodejs/loaders/issues/147
Open issue for Node v21 for google https://github.com/googleapis/nodejs-storage/issues/2368

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -461,7 +484,7 @@ workflows:
         - not:
           equal: [ scheduled_pipeline, << pipeline.trigger_source >> ]
     jobs:
-      - "node-20"
+      - "node-21"
```
Comment (aryamohanan) [inline .circleci/config.yml:487]: note: added as part of testing. TBD later

File: .circleci/config.yml [hunk 2]
```diff
@@ -297,6 +297,29 @@ memcached: &memcached
...
+      - <<: *mssql
+      - <<: *rabbitmq
+      - <<: *nats
+      - <<: *nats-streaming
+      - <<: *memcached
+      - <<: *couchbase
+      - <<: *couchbase-setup
+      - <<: *localstack
+    <<: *shared
+    parallelism: 15
```
[the remark this answers is not in the document]
Comment (aryamohanan) [reply .circleci/config.yml]: Updated

File: packages/autoprofile/precompile/build-all-addons.js [hunk 3]
```diff
@@ -32,12 +32,11 @@ const LIBC_FAMILIES = [GLIBC, MUSL];
 // Maintenance Note:
 // This should be kept in sync with native-dep-packs/rebuild-precompiled-addons.sh -> ABI_VERSIONS.
 const ABI_VERSIONS = {
-  // 14.17.x was no longer working
-  // https://stackoverflow.com/questions/76094428/debian-stretch-repositories-404-not-found
```
Comment (kirrg001) [inline packages/autoprofile/precompile/build-all-addons.js:36]: Question: Why did you remove the comments?

File: packages/autoprofile/precompile/docker-opts.js [hunk 4]
```diff
@@ -16,10 +16,7 @@ module.exports = exports = function getDockerOpts(family, version) {
   const targetNodeVersion = version || nodeJsVersion;
 
   if (family === GLIBC) {
-    // NOTE: debian stretch was no longer working good with some Node versions
-    // https://stackoverflow.com/questions/76094428/debian-stretch-repositories-404-not-found
-    // https://stackoverflow.com/questions/52083380/in-docker-image-names-what-is-the-difference-between-alpine-jessie-stretch-an
-    baseImage = `node:${targetNodeVersion}-buster`;
+    baseImage = `node:${targetNodeVersion}`;
```
Comment (kirrg001) [inline packages/autoprofile/precompile/docker-opts.js:19]: Question: Why did you remove buster postfix?

File: packages/collector/test/tracing/cloud/gcp/storage/test.js [hunk 5]
```diff
@@ -49,11 +50,14 @@ if (
   });
 } else {
   let mochaSuiteFn;
+  // There is bug in node 21.2.0, more details please see
+  // https://github.com/googleapis/nodejs-storage/issues/2368
 
-  if (!supportedVersion(process.versions.node) || !process.env.GCP_PROJECT) {
+  if (!supportedVersion(process.versions.node) || !process.env.GCP_PROJECT ||
+  semver.eq(process.versions.node, '21.2.0')) {
```
Comment (kirrg001) [inline packages/collector/test/tracing/cloud/gcp/storage/test.js:57]: I think this must check for >=?
Pls see referenced issue again :)

File: packages/autoprofile/precompile/docker-opts.js (hunk 4 above)
Comment (basti1302) [reply packages/autoprofile/precompile/docker-opts.js]: See commit comment:

> Also, use the Docker image node:${targetNodeVersion}
instead of node:${targetNodeVersion}-buster for building the autoprofile
binaries. Specifying a specific Debian distribution has a tiny
maintenance disadvantage, as node images are not always provided for all
Debian distributions. There is no node:21.2.0-buster, for example, as
Debian Buster is simply too old by now.

But if you want to bring them back, that's fine as well.

Comment (basti1302) [reply packages/autoprofile/precompile/docker-opts.js]: My thinking was: The next time you folks update for, let's say Node.js 26, and the Node.js project has stopped providing a node.js-bullseye image, you need to update this again. I thought we might not need to be that specific and just using nodes image without postfix is fine as well.

File: packages/collector/test/tracing/cloud/gcp/storage/test.js (hunk 5 above)
Comment (aryamohanan) [reply packages/collector/test/tracing/cloud/gcp/storage/test.js]: nice catch,

File: packages/autoprofile/precompile/build-all-addons.js (hunk 3 above)
Comment (basti1302) [reply packages/autoprofile/precompile/build-all-addons.js:36]: @kirrg001 ✋  I removed those. I felt they are no longer relevant. Seemed to more an explanation of why it was changed from 14.17. to 14.21. But 14.21. is the most recent 14.x version anyway, we wouldn't randomly go back to 14.17 anyway would we?

Comment (kirrg001) [reply packages/autoprofile/precompile/build-all-addons.js:36]: Oke 🫡

File: packages/autoprofile/precompile/docker-opts.js (hunk 4 above)
Comment (kirrg001) [reply packages/autoprofile/precompile/docker-opts.js:19]: I have added a TODO to #884.

File: .circleci/config.yml (hunk 1 above)
Comment (aryamohanan) [reply .circleci/config.yml:487]: Discussed. And we will keep the development version here (18)

**Review by kirrg001 — APPROVED** [review]

File: packages/collector/test/tracing/cloud/gcp/storage/test.js [hunk 6]
```diff
@@ -49,11 +50,14 @@ if (
   });
 } else {
   let mochaSuiteFn;
+  // There is bug in node 21.2.0, more details please see
+  // https://github.com/googleapis/nodejs-storage/issues/2368
 
-  if (!supportedVersion(process.versions.node) || !process.env.GCP_PROJECT) {
+  if (!supportedVersion(process.versions.node) || !process.env.GCP_PROJECT ||
+  semver.gte(process.versions.node, '21.2.0')) {
```
Comment (kirrg001) [inline packages/collector/test/tracing/cloud/gcp/storage/test.js:57]: @instana/eng-nodejs I have created a ticket for this. The bug is resolved. Good first task
